src/can.rs

```diff
@@ -0,0 +1,1060 @@
+            _ => self.0.cmp(&other.0),
+        }
+    }
+}
+
+impl PartialOrd for IdReg {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// A CAN data or remote frame.
+#[derive(Clone, Debug)]
+pub struct Frame {
+    id: IdReg,
+    dlc: usize,
+    data: [u8; 8],
+}
+
+impl Frame {
+    /// Creates a new frame.
+    pub fn new(id: Id, data: &[u8]) -> Result<Frame, ()> {
+        if !id.valid() || data.len() > 8 {
+            return Err(());
+        }
+
+        let id = match id {
+            Id::Standard(id) => IdReg::new_standard(id),
+            Id::Extended(id) => IdReg::new_extended(id),
+        };
+
+        let mut frame = Self {
+            id,
+            dlc: data.len(),
+            data: [0; 8],
+        };
+        frame.data[0..data.len()].copy_from_slice(data);
+        Ok(frame)
+    }
+
+    /// Creates a new remote frame with configurable data length code (DLC).
+    pub fn new_remote(id: Id, dlc: usize) -> Result<Frame, ()> {
+        if dlc >= 8 {
+            return Err(());
+        }
+
+        let mut frame = Self::new(id, &[])?;
+        frame.dlc = dlc;
+        frame.id.with_rtr(true);
+        Ok(frame)
+    }
+
+    /// Returns true if this frame is an extended frame
+    pub fn is_extended(&self) -> bool {
+        self.id.is_extended()
+    }
+
+    /// Returns true if this frame is a standard frame
+    pub fn is_standard(&self) -> bool {
+        self.id.is_standard()
+    }
+
+    /// Returns true if this frame is a remote frame
+    pub fn is_remote_frame(&self) -> bool {
+        self.id.rtr()
+    }
+
+    /// Returns true if this frame is a data frame
+    pub fn is_data_frame(&self) -> bool {
+        !self.is_remote_frame()
+    }
+
+    /// Returns the frame identifier.
+    pub fn id(&self) -> Id {
+        self.id.to_id()
+    }
+
+    /// Returns the data length code (DLC) which is in the range 0..8.
+    ///
+    /// For data frames the DLC value always matches the length of the data.
+    /// Remote frames no not carry any data, yet the DLC can be greater than 0.
+    pub fn dlc(&self) -> usize {
+        self.dlc
+    }
+
+    /// Returns the frame data (0..8 bytes in length).
+    pub fn data(&self) -> &[u8] {
+        if self.is_data_frame() {
+            &self.data[0..self.dlc]
+        } else {
+            &[]
+        }
+    }
+}
+
+// Ordering is based on the ID and can be used to sort frames by priority.
+impl Ord for Frame {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.id.cmp(&other.id)
+    }
+}
+
+impl PartialOrd for Frame {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+// The Equality traits compare the identifier and the data.
+impl PartialEq for Frame {
+    fn eq(&self, other: &Self) -> bool {
+        self.id() == other.id() && self.data[0..self.dlc] == other.data[0..other.dlc]
```

Comment:
The `PartialOrd` and `PartialEq` implementations of a type must agree with each other. With this implementation that is not the case, since two data frames with the same ID but different data would compare equal via `PartialOrd`, but not via `PartialEq`.